### PR TITLE
*: refactor tiflow-master/tiflow-executor logic of scaler

### PR DESCRIPTION
**Pull request by arlottang**
### What problem does this PR solve?
- Code refactor for Scaler

Issue link:

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Check List  

Tests 

- [ ] Unit test
- [ ] E2E test
- [ ] Stability test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Breaking backward compatibility
- [ ] Other side effects:

### CODE REVIEW

File: pkg/manager/member/master_scaler.go [hunk 1]
```diff
@@ -58,31 +103,97 @@ func (s masterScaler) ScaleIn(meta metav1.Object, actual *apps.StatefulSet, desi
...
+
+		if err := s.EvictLeader(tc, current-1); err != nil {
+			return nil
+		}
+
+		if err := s.SetReplicas(ctx, actual, uint(current-1)); err != nil {
+			return err
+		}
+
+		if err := s.WaitUntilHealthy(ctx, uint(current-1)); err != nil {
```
Comment (arlottang) [inline pkg/manager/member/master_scaler.go:129]: logic need to be changed

Comment (arlottang) [reply pkg/manager/member/master_scaler.go:129]: I think, If we have migrated the leader, we should make sure that the current leader is the correct one.

File: pkg/manager/member/executor_manager.go [hunk 2]
```diff
@@ -45,49 +44,45 @@ const (
...
 }
 
 // Sync implements the logic for syncing tiflowCluster executor member.
 func (m *executorMemberManager) Sync(ctx context.Context, tc *v1alpha1.TiflowCluster) error {
-
 	ns := tc.GetNamespace()
 	tcName := tc.GetName()
 
-	klog.Infof("start to sync tiflow cluster [%s/%s]", ns, tcName)
+	klog.Infof("start to sync tiflow-Executor [%s/%s]", ns, tcName)
```
Comment (lichunzhu) [inline pkg/manager/member/executor_manager.go]: maybe `tiflow executor`?

File: pkg/manager/member/master_scaler.go [hunk 3]
```diff
@@ -58,31 +103,97 @@ func (s masterScaler) ScaleIn(meta metav1.Object, actual *apps.StatefulSet, desi
...
+	down := *actual.Spec.Replicas - *desired.Spec.Replicas
+	current := *actual.Spec.Replicas
+	ctx := context.TODO()
+
+	for i := down; i > 0; i-- {
+		klog.Infof("scaling down statefulSet  %s of master, current: %d, desired: %d",
+			stsName, current, current-1)
+
+		if err := s.EvictLeader(tc, current-1); err != nil {
+			return nil
```
Comment (lichunzhu) [inline pkg/manager/member/master_scaler.go]: Why do we `return nil` here?

File: pkg/manager/member/master_scaler.go [hunk 4]
```diff
@@ -58,31 +100,97 @@ func (s masterScaler) ScaleIn(meta metav1.Object, actual *apps.StatefulSet, desi
...
+		stsName, ns, tcName, *actual.Spec.Replicas, *desired.Spec.Replicas)
 
-	if !tc.Status.Master.Synced {
-		return fmt.Errorf("tiflow cluster: %s/%s's tiflow-master status sync failed, can't scale in now", ns, tcName)
+	down := *actual.Spec.Replicas - *desired.Spec.Replicas
+	current := *actual.Spec.Replicas
+	ctx := context.TODO()
+
+	for i := down; i > 0; i-- {
+		klog.Infof("scaling down statefulSet  %s of master, current: %d, desired: %d",
```
Comment (lichunzhu) [inline pkg/manager/member/master_scaler.go]:
```suggestion
		klog.Infof("scaling down statefulSet %s of master, current: %d, desired: %d",
```

**Review by lichunzhu — APPROVED** [review]

**Pull request merged**